lib/std/compress/zstandard.zig

```diff
@@ -0,0 +1,286 @@
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+const RingBuffer = std.RingBuffer;
+
+const types = @import("zstandard/types.zig");
+pub const frame = types.frame;
+pub const compressed_block = types.compressed_block;
+
+pub const decompress = @import("zstandard/decompress.zig");
+
+pub const DecompressStreamOptions = struct {
+    verify_checksum: bool = true,
+    window_size_max: usize = 1 << 23, // 8MiB default maximum window size,
+};
+
+pub fn DecompressStream(
+    comptime ReaderType: type,
+    comptime options: DecompressStreamOptions,
+) type {
```

Comment:
The `gzip`, `lzma`, and `xz` implementations all expose the same `std.compress.foo.Decompress` signature:
```zig
pub fn Decompress(comptime ReaderType: type) type {
```
This is leveraged in commit d94613c1d06fbeaf0cee88a04842cda64a10c8f9 for example to make some code generic over the decompression algorithm.

Is it strictly necessary to have the options comptime-known for the zstd implementation? If not I think it would be better to conform to the existing interface and instead pass the options at runtime to the init() function. This would also reduce generic code bloat in some cases.

Furthermore, `std.compress.zstd` feels more consistent with the current members of the `std.compress` namespace to me:
```zig
pub const deflate = @import("compress/deflate.zig");
pub const gzip = @import("compress/gzip.zig");
pub const lzma = @import("compress/lzma.zig");
pub const lzma2 = @import("compress/lzma2.zig");
pub const xz = @import("compress/xz.zig");
pub const zlib = @import("compress/zlib.zig");
```

Reply:
I agree that `std.compress.zstd` is probably a better name to go with. As for the `std.compress.foo.Decompress` signature, the package manager is only using the `std.compress.foo.decompress()` function which isn't affected by the comptime parameters, but will probably require at least the `window_size_max` comptime parameter to `DecompressStream` to match this API.

It is not strictly necessary to have the options be comptime - the `verify_checksum` option being comptime just allows avoiding, but the main thing is that not having them be comptime will either require us to hard code the values (for `max_window_size` this is not a good solution as someone  may have Zstandard frames they cannot decode because of this), or some other API will have to diverge from the others in `std.compress`, most likely the `init()` function as you mentioned, but this is a more intrusive divergence from the other decompression algorithms which I don't think will really allow for code that is generic across the different decompression algorithms. In normal usage you can just use the default comptime parameters to get an API that matches that of the other decompression algorithms.

I have a branch that adds Zstandard support to the package manager generically just like gzip and xz - the first commit on the branch renames all the `Decompress`/`decompress` functions to `DecompressStream`/`decompressStream` which I think is a better name. As I mentioned earlier if we really want to match the APIs we may need a separate issue to decide on one that works well for all of them (e.g. the Zstandard `init()` function has no need to return an error union, which the others currently do, though I don't know how necessary this is) and I don't think `Decompress`/`decompress` is a great name, especially when you have non-streaming decompression functions available.

Reply:
FWIW, this is how the lzma decompressor currently handles options at runtime:
```zig
pub const decode = @import("lzma/decode.zig");

pub fn decompress(
    allocator: Allocator,
    reader: anytype,
) !Decompress(@TypeOf(reader)) {
    return decompressWithOptions(allocator, reader, .{});
}

pub fn decompressWithOptions(
    allocator: Allocator,
    reader: anytype,
    options: decode.Options,
) !Decompress(@TypeOf(reader)) {
    const params = try decode.Params.readHeader(reader, options);
    return Decompress(@TypeOf(reader)).init(allocator, reader, params, options.memlimit);
}

pub fn Decompress(comptime ReaderType: type) type {
     ...
```
It seems to me that the zstd implementation could use the same pattern here if I understand correctly that the options aren't strictly required to be comptime-known.